modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/ags_linear_solver.cc

```diff
@@ -16,107 +17,77 @@ namespace opensn
 namespace lbs
 {
 
-void
-AGSLinearSolver::SetSystemSize()
-{
-  CALI_CXX_MARK_SCOPE("AGSLinearSolver::SetSystemSize");
-
-  auto ags_context_ptr = std::dynamic_pointer_cast<AGSContext>(context_ptr_);
-
-  const auto sizes = ags_context_ptr->SystemSize();
-
-  num_local_dofs_ = sizes.first;
-  num_global_dofs_ = sizes.second;
-}
-
-void
-AGSLinearSolver::SetSystem()
-{
-  CALI_CXX_MARK_SCOPE("AGSLinearSolver::SetSystem");
-
-  x_ = CreateVector(static_cast<int64_t>(num_local_dofs_), static_cast<int64_t>(num_global_dofs_));
-
-  VecSet(x_, 0.0);
-  VecDuplicate(x_, &b_);
-
-  // Create the matrix-shell
-  MatCreateShell(opensn::mpi_comm,
-                 static_cast<int64_t>(num_local_dofs_),
-                 static_cast<int64_t>(num_local_dofs_),
-                 static_cast<int64_t>(num_global_dofs_),
-                 static_cast<int64_t>(num_global_dofs_),
-                 &(*context_ptr_),
-                 &A_);
-
-  // Set the action-operator
-  MatShellSetOperation(A_, MATOP_MULT, (void (*)())LinearSolverMatrixAction);
-
-  // Set solver operators
-  KSPSetOperators(ksp_, A_, A_);
-  KSPSetUp(ksp_);
-}
-
-void
-AGSLinearSolver::SetPreconditioner()
-{
-  CALI_CXX_MARK_SCOPE("AGSLinearSolver::SetPreconditioner");
-
-  auto ags_context_ptr = std::dynamic_pointer_cast<AGSContext>(context_ptr_);
-  ags_context_ptr->SetPreconditioner(ksp_);
-}
-
-void
-AGSLinearSolver::SetRHS()
-{
-}
-
-void
-AGSLinearSolver::SetInitialGuess()
-{
-}
-
 void
 AGSLinearSolver::Solve()
 {
   CALI_CXX_MARK_SCOPE("AGSLinearSolver::Solve");
 
   auto ags_context_ptr = std::dynamic_pointer_cast<AGSContext>(context_ptr_);
+  auto grid_ptr = GetCurrentMesh();
   auto& lbs_solver = ags_context_ptr->lbs_solver_;
-
-  const int gid_i = GroupSpanFirstID();
-  const int gid_f = GroupSpanLastID();
-  const auto& phi = lbs_solver.PhiOldLocal();
-
-  Vec x_old;
-  VecDuplicate(x_, &x_old);
-
-  // Save qmoms to be restored after each iteration.
-  // This is necessary for multiple ags iterations to function
-  // and for keigen-value problems
+  auto& wgs_solvers = ags_context_ptr->wgs_solvers_;
+  auto& cell_transport_views = lbs_solver.GetCellTransportViews();
+  auto& groupsets = lbs_solver.Groupsets();
+  auto& phi_new = lbs_solver.PhiNewLocal();
+  auto num_moments = lbs_solver.NumMoments();
+  std::vector<double> phi_old(lbs_solver.PhiOldLocal().size(), 0.0);
+
+  auto piecewise_change =
```

Comment:
This feels like it should live somewhere else to me

Reply:
I could make it a separate method in `AGSLinearSolver`, but I end up having to do a bunch of variable assignments and things that had already been done in `Solve`. In Chi-Tech, a very similar method lived in `LBSSolver` and was used for the Classic Richardson WGS routine. I could move it there. To be honest, I'm not a fan of any of the solutions. Maybe a separate routine in `AGSLinearSolver` is best?

Reply:
Hmm... Well here's my first question: Where will something like this ultimately be used? It feels like it should be available to the `WGSLinearSolver` as well, doesn't it? If yes, I'd say that `LBSSolver` makes the most sense.

As a side note, the loop over groupsets is not entirely necessary here. Flux moments are stored for all groups so this could be a single loop over `g < lbs_solver.NumGroups()` instead.

Reply:
Fixed. I made the pointwise check a private member of the AGSSolver class rather than move it into LBSSolver. It's unique in the fact that it computes the pointwise change between LSBSolver::PhiNewLocal() and a private AGSSolver member variable, so that seemed the appropriate place.